### ISSUE TITLE
chop off the end symbol in decoder?

### ISSUE THREAD

**Issue by hzxscyq** [Luvata/CS224N-2019#5]
In nmt_model.py line 215, the code chops off the last row of the target padded. What does this mean? It just deletes the END token of the longest sentences and other END tokens of the shorter sentences are kept?  I would be very appreciate for your help.

**Comment by Luvata** [comment]
Sorry for abandon this project that long
@hzxscyq That's a good question, because the naming of variables are a little confused, you'll probably missed that.
In pytorch, the output of LSTM has shape `(seq_len, batch, num_directions * hidden_size)` [doc](https://pytorch.org/docs/stable/generated/torch.nn.LSTM.html). So the last row that was chops off from `target_padded[:-1]` is the last tokens (which is `<eos>`).
In this assignment, [Teacher forcing](https://en.m.wikipedia.org/wiki/Teacher_forcing) is using for training. In short, for a target sentence `<bos> Good morning <eos>`, teaching forcing will calculate loss for each pair of tokens
```
input:       <bos>         Good         morning 
target:      Good         morning       <eos>
``` 
The `input` is the "chops off of the last tokens" that you mention in line [215](https://github.com/Luvata/CS224N-2019/blob/master/Assignment/a4/nmt_model.py#L215). And the `target` is also the "chops off of the first tokens" in line [121](https://github.com/Luvata/CS224N-2019/blob/master/Assignment/a4/nmt_model.py#L121).

**Comment by Luvata** [comment]
Furthermore, for a batch of sentences with different lengths, the last tokens of `padded_sequences` will be the `eos` of the longest sentence as you mentioned above. 

For a shorter sentence in this batch, the `eos` will not be chops off, and since the `<PAD>` tokens was applied to the right, the target of that `eos` token is `<PAD>` token. But since the loss is not calculated in `<PAD>` token, the teacher forcing still worked correctly.

That's a really interesting question !
